refactor: Make BitcoinUnits::Unit a scoped enum

hebasto opened the pull request:
Ported from https://github.com/bitcoin/bitcoin/pull/17877.

Since Qt 5.5 there are [means](https://doc.qt.io/qt-5/qobject.html#Q_ENUM) to register an enum type with the meta-object system (such enum still lacks an ability to interact with [`QSettings::setValue()`](https://doc.qt.io/qt-5/qsettings.html#setValue) and [`QSettings::value()`](https://doc.qt.io/qt-5/qsettings.html#value) without defined stream operators).

In order to reduce global namespace polluting and to force strong type checking, this PR makes `BitcoinUnits::Unit` a scoped enum (`typedef BitcoinUnits::Unit BitcoinUnit;`).

No behavior change.

hebasto commented:
bitcoin/bitcoin#17877 has Concept ACKs from:
- @promag 
- @practicalswift 
- @elichai 
- @jonasschnelli 
- @Empact 

---

@promag @jonasschnelli Mind reviewing this PR?

hebasto commented:
Rebased 243205f85b8b10e4f43865557cdd8ea3aa71c380 -> b109ae6d2b292f561d2afbc3e4985651f06575be ([pr60.01](https://github.com/hebasto/gui/commits/pr60.01) -> [pr60.02](https://github.com/hebasto/gui/commits/pr60.02)) due to the conflict with #35.

Sjors commented:
tACK b109ae6

I just noticed one practical downside of this separate repo; having to add another remote: `git remote add hebasto-gui git@github.com:hebasto/gui.git`

promag commented:
@Sjors I only have git@github.com:bitcoin-core/gui.git and my gui fork.

MarcoFalke commented:
I manually call `git fetch https://github.com/hebasto/gui/ 200814-unit`, which works without a remote

elichai commented:
FWIW, I'm like @promag and using @jonatack's scripts to fetch all PRs:
```
[remote "upstream"]
        fetch = +refs/pull/*/head:refs/remotes/upstream/pr/*

```

then you can just `git checkout upstream/pr/60.`

hebasto commented:
Rebased b109ae6d2b292f561d2afbc3e4985651f06575be -> 7e636c7fe129afff8ff248cdd6b55e59445f1e0e ([pr60.02](https://github.com/hebasto/gui/commits/pr60.02) -> [pr60.03](https://github.com/hebasto/gui/commits/pr60.03)) due to the conflict with #176.

jonatack commented:
Concept ACK

promag commented on src/qt/bitcoinunits.cpp:
```diff
@@ -248,3 +248,39 @@ CAmount BitcoinUnits::maxMoney()
+    case 3: return BitcoinUnits::SAT;
+    case 2: return BitcoinUnits::uBTC;
+    case 1: return BitcoinUnits::mBTC;
+    default: return BitcoinUnits::BTC;
```
77edf1bd0149bd2625dff563f3bd4032a12a2233

Add case for `0` and an `assert(false)` in default?

hebasto replied:
Thanks! [Updated](https://github.com/bitcoin-core/gui/pull/60#issuecomment-823554773).

promag commented on src/qt/bitcoinunits.cpp:
```diff
@@ -248,3 +248,39 @@ CAmount BitcoinUnits::maxMoney()
 }
+
+namespace {
+qint8 ToQuint8(BitcoinUnit unit)
```
77edf1bd0149bd2625dff563f3bd4032a12a2233

ToQint8?

hebasto replied:
Thanks! [Updated](https://github.com/bitcoin-core/gui/pull/60#issuecomment-823554773).

promag commented on src/qt/optionsmodel.cpp:
```diff
@@ -70,9 +71,15 @@ void OptionsModel::Init(bool resetSettings)
 
     // Display
-    if (!settings.contains("nDisplayUnit"))
-        settings.setValue("nDisplayUnit", BitcoinUnits::BTC);
```
77edf1bd0149bd2625dff563f3bd4032a12a2233

Maybe it should migrate the old setting key if it exists?

hebasto replied:
Will it be dead code after a while?

promag replied:
And? For instance https://github.com/bitcoin/bitcoin/pull/20966 also tries to use old data.

hebasto replied:
@promag 
I was thinking about your suggestion, and I failed to find a safe and short (about 1..2 lines of code) solution.
OTOH, adjusting this option requires only two clicks.

Therefore, I'm going to leave it as is, leaving a room for improvements in follow ups.

promag commented on src/qt/optionsmodel.cpp:
```diff
@@ -501,12 +508,11 @@ bool OptionsModel::setData(const QModelIndex & index, const QVariant & value, in
 {
-    if (!value.isNull())
-    {
+    if (!value.isNull()) {
```
77edf1bd0149bd2625dff563f3bd4032a12a2233

nit, could skip if `value == m_display_unit`?

hebasto replied:
Thanks! [Updated](https://github.com/bitcoin-core/gui/pull/60#issuecomment-823554773).

promag reviewed (commented):
Code review ACK 7e636c7fe129afff8ff248cdd6b55e59445f1e0e with some comments.

Will test and play around.

hebasto commented:
Updated 7e636c7fe129afff8ff248cdd6b55e59445f1e0e -> cb08a1c22d5bbf7baf5dc5003dd29185660efd78 ([pr60.03](https://github.com/hebasto/gui/commits/pr60.03) -> [pr60.04](https://github.com/hebasto/gui/commits/pr60.04), [diff](https://github.com/hebasto/gui/compare/pr60.03..pr60.04)):

- addressed most of @promag's comments

hebasto commented:
Updated cb08a1c22d5bbf7baf5dc5003dd29185660efd78 -> 68baad3cfac5c0daf9589a92ea1e99c20cc0e380 ([pr60.04](https://github.com/hebasto/gui/commits/pr60.04) -> [pr60.05](https://github.com/hebasto/gui/commits/pr60.05)):

- rebased due to the conflict with #263
- option renamed s/`display_unit`/`DisplayBitcoinUnit`/ to be in line with other recently added options, see #295

hebasto commented:
Rebased 68baad3cfac5c0daf9589a92ea1e99c20cc0e380 -> 8eb1a34937990012d661d262a029e33c8cb67776 ([pr60.05](https://github.com/hebasto/gui/commits/pr60.05) -> [pr60.06](https://github.com/hebasto/gui/commits/pr60.06)) due to the conflict with #275.

jb55 commented:
Code Review ACK 8eb1a34937990012d661d262a029e33c8cb67776

looks like a reasonable cleanup

jarolrod commented:
cc @hebasto, a rebase is needed :)

jb55 commented:
I have a rebased version here: https://github.com/bitcoin-core/gui/compare/master...jb55:units-scoped-enum

I can submit a PR if you want me to take this over @hebasto

hebasto commented:
> I have a rebased version here: [master...jb55:units-scoped-enum](https://github.com/bitcoin-core/gui/compare/master...jb55:units-scoped-enum)
> 
> I can submit a PR if you want me to take this over @hebasto

@jb55 

Thank you! I'll appreciate if you do that.

hebasto commented:
Closing in favor of #556.
